lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,37 +1175,134 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
+  llvm::SmallVector<llvm::MDNode *, 4> Entries;
+
   // Collect the linker options already in the module (from ClangCodeGen).
-  // FIXME: This constant should be vended by LLVM somewhere.
-  auto *Metadata = Module.getOrInsertNamedMetadata("llvm.linker.options");
   for (llvm::MDNode *LinkOption : Metadata->operands())
-    AutolinkEntries.push_back(LinkOption);
+    Entries.push_back(LinkOption);
+
+  llvm::LLVMContext &ctx = Module.getContext();
+
+  // Construct autolinking entries
+  switch (Kind) {
+  case AutolinkKind::LLVMLinkerOptions:
+  case AutolinkKind::SwiftAutoLinkExtract: {
+    // On platforms that support autolinking, continue to use the metadata.
+    for (LinkLibrary &linkLib : AutolinkEntries) {
+      switch (linkLib.getKind()) {
+      case LibraryKind::Library: {
+        llvm::SmallString<32> opt =
+            getTargetDependentLibraryOption(Triple, linkLib.getName());
+        Entries.push_back(
+            llvm::MDNode::get(ctx, llvm::MDString::get(ctx, opt)));
+        break;
+      }
+      case LibraryKind::Framework: {
+        llvm::Metadata *args[] = {llvm::MDString::get(ctx, "-framework"),
+                                  llvm::MDString::get(ctx, linkLib.getName())};
+        Entries.push_back(llvm::MDNode::get(ctx, args));
+        break;
+      }
+      }
+    }
+    break;
+  }
+  case AutolinkKind::LLVMDependentLibraries: {
+    for (LinkLibrary &linkLib : AutolinkEntries) {
+      switch (linkLib.getKind()) {
+      case LibraryKind::Library: {
+        Entries.push_back(llvm::MDNode::get(
+            ctx, llvm::MDString::get(ctx, linkLib.getName())));
+        break;
+      }
+      case LibraryKind::Framework: {
+        llvm_unreachable(
+            "llvm.dependent-libraries doesn't support framework dependency");
+        break;
+      }
+      }
+    }
+    break;
+  }
+  }
 
-  // Remove duplicates.
   llvm::SmallPtrSet<llvm::MDNode *, 4> knownAutolinkEntries;
-  AutolinkEntries.erase(std::remove_if(AutolinkEntries.begin(),
-                                       AutolinkEntries.end(),
-                                       [&](llvm::MDNode *entry) -> bool {
-                                         return !knownAutolinkEntries.insert(
-                                                   entry).second;
-                                       }),
-                        AutolinkEntries.end());
-
-  const bool AutolinkExtractRequired =
-      (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4()) ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::Wasm ||
-      Triple.isOSCygMing();
-
-  if (!AutolinkExtractRequired) {
+  Entries.erase(
+      std::remove_if(Entries.begin(), Entries.end(),
+                     [&](llvm::MDNode *entry) -> bool {
+                       return !knownAutolinkEntries.insert(entry).second;
+                     }),
+      Entries.end());
+
+  // Write the entries
+  switch (Kind) {
+  case AutolinkKind::LLVMLinkerOptions:
```

Comment:
For instance, if one did a struct enum like pattern, this could be refactored into a method on the struct enum.